junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertThrows.java

```diff
@@ -38,8 +42,37 @@
 				throw new AssertionFailedError(message, actualException);
 			}
 		}
-		throw new AssertionFailedError(
-			String.format("Expected %s to be thrown, but nothing was thrown.", getCanonicalName(expectedType)));
+		throw new AssertionFailedError(getFailureMessage(expectedType));
+	}
+
+	static <T extends Throwable> T assertThrows(Class<T> expectedType, Executable executable, String message) {
+		return assertThrows(expectedType, executable, () -> message);
 	}
 
+	@SuppressWarnings("unchecked")
+	static <T extends Throwable> T assertThrows(Class<T> expectedType, Executable executable,
+			Supplier<String> messageSupplier) {
+		try {
+			executable.execute();
+		}
+		catch (Throwable actualException) {
+			if (expectedType.isInstance(actualException)) {
+				return (T) actualException;
+			}
+			else {
+				throw new AssertionFailedError(getFailureWithSuppliedMessage(expectedType, messageSupplier),
+					actualException);
+			}
+		}
+		throw new AssertionFailedError(getFailureWithSuppliedMessage(expectedType, messageSupplier));
+	}
+
+	private static <T extends Throwable> String getFailureWithSuppliedMessage(Class<T> expectedType,
+			Supplier<String> messageSupplier) {
+		return buildPrefix(nullSafeGet(messageSupplier) + getFailureMessage(expectedType));
```

Comment:
The closing parenthesis for `buildPrefix()` must follow the call to `nullSafeGet()`.